### PR TITLE
✨ SwG crypto: decryption implementation

### PR DESCRIPTION
**Pull request by chenshay**
Implementing decryption for SwG Crypto

@ampproject/wg-access-subscriptions
#21658

### CODE REVIEW

File: build-system/app.js [hunk 1]
```diff
@@ -1342,4 +1337,18 @@ function generateInfo(filePath) {
       '<h3><a href = /serve_mode=cdn>Change to CDN mode (prod JS)</a></h3>';
 }
 
+function decryptDocumentKey(encryptedDocumentKey) {
+  const cryptoStart = 'ENCRYPT(';
+  if (!encryptedDocumentKey.includes(cryptoStart, 0)) {
```
Comment (dvoytenko) [inline build-system/app.js]: My guess is that `encryptedDocumentKey` in this context could be `null|undefined`, so this should probably be the first check.

File: extensions/amp-subscriptions/0.1/crypto-handler.js [hunk 2]
```diff
@@ -92,11 +92,36 @@ export class CryptoHandler {
...
    */
   decryptDocumentContent_(
-    // eslint-disable-next-line no-unused-vars
     encryptedContent, documentKey) {
-    // Don't really return this. Placeholder for the real thing.
-    // const placeholder = encryptedContent.trim() + documentKey;
-    const placeholder = '<h2><i> abc </i></h2>';
-    return Promise.resolve(placeholder);
+
+console.log('encryptedContent', encryptedContent);
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/crypto-handler.js]: a note not to forget to remove

File: extensions/amp-subscriptions/0.1/crypto-handler.js [hunk 3]
```diff
@@ -92,11 +92,36 @@ export class CryptoHandler {
...
+      return new TextDecoder().decode(new Uint8Array(buffer));
+    });
+    return Promise.resolve('<h2><i>' + decryptedContent + '</i></h2>');
+  }
+
+  /**
+   * @param {string} s
+   * @return {ArrayBuffer}
+   */
+  base64ToBytes(s) {
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/crypto-handler.js]: There's `base64.js` file. Let's use the utilities from there.

File: extensions/amp-subscriptions/0.1/crypto-handler.js [hunk 4]
```diff
@@ -92,11 +92,36 @@ export class CryptoHandler {
...
+          name: 'AES-CTR',
+          counter: new Uint8Array(16), // iv: all zeros.
+          length: 128, // block size (16): 1-128
+        },
+        this.base64ToBytes(documentKey),
+        encryptedContent.trim(),
+    ).then(function(buffer) {
+      return new TextDecoder().decode(new Uint8Array(buffer));
+    });
+    return Promise.resolve('<h2><i>' + decryptedContent + '</i></h2>');
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/crypto-handler.js]: Let's remove `<h2><i>` now. If you'd like, you can add it to the encrypted content itself.

File: extensions/amp-subscriptions/0.1/crypto-handler.js [hunk 5]
```diff
@@ -92,11 +92,36 @@ export class CryptoHandler {
...
+console.log('documentKey', documentKey);
+
+    const decryptedContent = crypto.subtle.decrypt(
+        {
+          name: 'AES-CTR',
+          counter: new Uint8Array(16), // iv: all zeros.
+          length: 128, // block size (16): 1-128
+        },
+        this.base64ToBytes(documentKey),
+        encryptedContent.trim(),
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/crypto-handler.js]: We can't just decrypt the text. First we need to extract bytes from it. Here's how we should approach it:

1. We need to trim and remove all whitespaces in between (e.g. line breaks). This way we get the complete encoded and encrypted string.
2. Un-base64 the text content. This way we get the actual crypted bytes.
3. Call `decrypt`.

File: build-system/app.js (hunk 1 above)
Comment (chenshay) [reply build-system/app.js]: done

File: extensions/amp-subscriptions/0.1/crypto-handler.js [hunk 6]
```diff
@@ -91,12 +93,51 @@ export class CryptoHandler {
...
-  decryptDocumentContent_(
-    // eslint-disable-next-line no-unused-vars
-    encryptedContent, documentKey) {
-    // Don't really return this. Placeholder for the real thing.
-    // const placeholder = encryptedContent.trim() + documentKey;
-    const placeholder = '<h2><i> abc </i></h2>';
-    return Promise.resolve(placeholder);
+  decryptDocumentContent_(encryptedContent, documentKey) {
+    // 1. Trim and remove all whitespaces (e.g. line breaks).
+    encryptedContent = encryptedContent.trim();
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/crypto-handler.js]: This trims, but doesn't remove whitespace.

File: extensions/amp-subscriptions/0.1/crypto-handler.js [hunk 7]
```diff
@@ -91,12 +93,51 @@ export class CryptoHandler {
...
+      const decryptedContent = crypto.subtle.decrypt(
+          {
+            name: 'AES-CTR',
+            counter: new Uint8Array(16), // iv: all zeros.
+            length: 128, // block size (16): 1-128
+          },
+          formattedDocKey,
+          encryptedBytes,
+      ).then(function(buffer) {
+        return base64EncodeFromBytes(buffer);
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/crypto-handler.js]: I'd not expect to have encode here. I'd expect UTF8 decoding though.

File: extensions/amp-subscriptions/0.1/crypto-handler.js [hunk 8]
```diff
@@ -91,12 +93,51 @@ export class CryptoHandler {
...
+    });
+  }
+
+  /**
+   * @private
+   * @param {string} documentKey
+   * @return {Promise<CryptoKey>}
+   */
+  stringToCryptoKey_(documentKey) {
+    // 1. Trim and remove all whitespaces (e.g. line breaks).
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/crypto-handler.js]: We shouldn't trim/clean this value - this is a precise key encoding.

File: extensions/amp-subscriptions/0.1/crypto-handler.js [hunk 9]
```diff
@@ -94,15 +96,47 @@ export class CryptoHandler {
...
+          return utf8Decode(new Uint8Array(buffer));
+        });
+      return decryptedContent;
+    });
+  }
+
+  /**
+   * @private
+   * @param {string} documentKey
+   * @return {Promise<!webCrypto.CryptoKey>}
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/crypto-handler.js]: `!`

File: extensions/amp-subscriptions/0.1/test/test-crypto-handler.js [hunk 10]
```diff
@@ -100,20 +104,28 @@ describes.realWin(
...
     describe('tryToDecryptDocument', () => {
       // eslint-disable-next-line max-len
       it('should replace the encrypted content with decrypted content in multiple sections', () => {
-        return cryptoHandler
-          .tryToDecryptDocument('decryptedDocumentKey')
-          .then(() => {
-            expect(cryptDiv1.textContent).to.equal(' abc ');
-            expect(cryptDiv2.textContent).to.equal(' abc ');
-          });
+        return cryptoHandler.tryToDecryptDocument(decryptedDocKey).then(() => {
```
Comment (dvoytenko) [inline extensions/amp-subscriptions/0.1/test/test-crypto-handler.js]: Can we also add a test where we stub the `Subtle.decrypt` method? Some specific arguments are important, such as `iv`, which may not always come through in a pre-canned test.

**Pull request merged**